Make ArrayOverBuffer behave like an Array/Array.sort no longer mutates the Array

hubertp opened the pull request:
### Pull Request Description

Most of the problems with accessing `ArrayOverBuffer` have been resolved by using `CoerceArrayNode` (https://github.com/enso-org/enso/pull/3817). In `Array.sort` we still however specialized on Array which wasn't compatible with `ArrayOverBuffer`. Similarly sorting JS or Python arrays wouldn't work.

Added a specialization to `Array.sort` to deal with that case. A generic specialization (with `hasArrayElements`) not only handles `ArrayOverBuffer` but also polyglot arrays coming from JS or Python. We could have an additional specialization for `ArrayOverBuffer` only (removed in the last commit) that returns `ArrayOverBuffer` rather than `Array` although that adds additional complexity which so far is unnecessary.

Also fixed an example in `Array.enso` by providing a default argument.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

hubertp committed:
Make ArrayOverBuffer behave like an Array

Most of the problems with accessing `ArrayOverBuffer` have been resolved
by using `CoerceArrayNode` (https://github.com/enso-org/enso/pull/3817).
In `Array.sort` we still however specialized on Array which wasn't
compatible with `ArrayOverBuffer`.

Added a specialization to `Array.sort` to deal with that case. Because
one cannot use a custom comparator with primitive (Java) arrays there is
a conversion needed. It's a necessary penalty if we want to keep
`ArrayOverBuffer` around.

Also fixed an example in `Array.enso` by providing a default argument.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayOverBuffer.java:
```diff
@@ -38,4 +50,49 @@ long getArraySize() {
+      throws UnsupportedMessageException {
+    if (child == null) {
+      return "null";
+    } else if (child instanceof Boolean) {
+      return (boolean) child ? "True" : "False";
+    } else {
```
I'm really worried that this logic for converting true/false to their Enso-textual representation is scattered throughout the project. It's in 3 places already and this instance is fourth.

I know that this logic is unlikely to change, so it's probably not a problem, but just so we can locate that - can we at least fold the `"True"` and `"False"` into some constants?

Still, ideally I think we should encapsulate this whole `showObject` logic (i.e. the part checking for `null`, for Boolean and then delegating to `iop`) into a separate function. It seems really unclean to repeat it over and over again.

radeusgd replied:
Other question is - in this particular scenario - is this code even needed?

I'd find it acceptable to just delegate to a common 'handle showObject' code and not think about particular cases as it makes the code simplest to reason about.

But here I think we have code visualizing a _byte_ array, which is written to handle booleans. I don't think we can have booleans in the byte buffer.

Still, I think it's better to just move this into a common function and call that one in all cases (handling booleans). But if we were to not do that, let's at least remove the special case for booleans which is in practice dead code here.

hubertp replied:
True, we can simplify it further. Let me take a look.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayOverBuffer.java:
```diff
@@ -38,4 +50,49 @@ long getArraySize() {
   }
+
+  @ExportMessage
+  String toDisplayString(boolean allowSideEffects) {
```
As below, this is basically copied over from `Vector`. Cannot we have this common logic extracted into some common function and delegated to, instead of duplicating the code? It's quite simple but it's a big chunk of logic here, so duplicating it just seems bad.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java:
```diff
@@ -52,19 +67,30 @@ Object doOther(State state, Object self, Object comparator) {
   }
 
+  Object runSort(Comparator<Object> compare, byte[] self, EnsoContext context) {
+    Byte[] tmpArray = ArrayUtils.toObject(self);
```
> Because one cannot use a custom comparator with primitive (Java) arrays there is a conversion needed.

This kind of conversion isn't going to be fast. I doubt the compiler will be able to virtualize away the `Byte[]` allocation. Anyway the sorting is done by `doSort` method which is behind `@TruffleBoundary`, so we can just dream of PE performance. Moreover we are unlikely to do anything better in this case (unless we want to write our own sort). 

PS: I was thinking of creating a _virtualized_ `java.util.List` delegating to the byte buffer and then just calling `List.sort()`, but the implementation in the `default List.sort` isn't really helping us anyway. The first thing it does is `toArray()` conversion.

hubertp replied:
Yes, performance wasn't the biggest priority at this point, compatibility was.
So far we don't have any other use-cases where this is really needed so I can file a ticket to investigate it although it won't be a priority.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java:
```diff
@@ -34,15 +42,22 @@ static SortNode build() {
   }
 
+  @Specialization
+  Object doArrayOverBuffer(State state, ArrayOverBuffer self, Function comparator) {
```
Can we sort JavaScript array? I'd expect specialization to `iop.hasArrayElements(self)`, but I don't see it anywhere.

hubertp replied:
Now we can (see next specialization). However I'm now wondering if it is worth keeping this one since the next one would also handle `ArrayOverBuffer`. Except it would return sorted `Array` rather than `ArrayOverBuffer`, so it really depends how important that is.

JaroslavTulach replied:
Exactly. `ArrayOverBuffer` is just an `iop.hasArrayElements(...)` and as such it could be handled by the next specialization in a generic way.

JaroslavTulach reviewed (approved):
I don't like `Array.sort` as it breaks referential transparency. But we have got it - e.g. good job on making the `sort` work on various types of arrays.

hubertp committed:
Make Array.sort not sorting in place

Sorting in place would be impractical to support for Arrays coming from
JS or Python.
That is why `Array.sort` now returns a copy of the original Array(-like)
structure, sorted.
This eliminates the need for createing a copy of Array for
`Vector.sort`.

Applied DRY to displaying Array-like structures.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/DisplayArrayUtils.java:
```diff
@@ -0,0 +1,50 @@
+
+public class DisplayArrayUtils {
+
+  public static String toDisplayString(
```
@JaroslavTulach for now I refactored the common code to this static method. I do wonder if it is really worth making this a Truffle Node.

JaroslavTulach replied:
If the code is behind `@TruffleBoundary` then it doesn't make any sense to make it a `Node`. The code concatenates strings - e.g. it has to be behind boundary. E.g. static method is completely fine.

hubertp committed:
More tests

hubertp committed:
Add missing TruffleBoundaries

jdunkerley commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -134,19 +134,22 @@ type Array
 
-       Arguments:
-       - comparator: A comparison function that takes two elements and returns
-         an Ordering that describes how the first element is ordered with
-         respect to the second.
+      Sorts the Array.
```
No longer needs to be ADVANCED as returning a new instance.

jdunkerley commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java:
```diff
@@ -2,28 +2,46 @@
 import java.util.Arrays;
 import java.util.Comparator;
+
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.apache.commons.lang3.ArrayUtils;
```
minor thing but can we tidy the imports here.

hubertp committed:
Simplify implementation

Skip ArrayOverBuffer specialization as it is now handled by
a generic `hasArrayElements` one.

hubertp committed:
update changelog

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java:
```diff
@@ -2,24 +2,35 @@
+@BuiltinMethod(
+    type = "Array",
+    name = "sort_builtin",
+    description = "Sorts a mutable array in place.")
```
The description needs to be updated to reflect the current behaviour.

```suggestion
    description = "Returns a sorted array.")
```

radeusgd commented on test/Tests/src/Semantic/Python_Interop_Spec.enso:
```diff
@@ -87,7 +90,18 @@ spec =
-            vec.map .x . should_equal [10, 20, 30]
+            vec.map .x . should_equal [30, 10, 20]
+
+            arr = vec.map .x . to_array
+            sorted = arr.sort
```
This is not a Python array anymore, right? After a `Vector.map` operation, this will be an Enso allocated Java array, no?

Still, a good test but this part is not really testing what it says - only the part below is _actually_ invoking our sort on a Python array instance.

hubertp replied:
Indeed, yes.

radeusgd reviewed (approved):
Looks good, just the comment should be updated before the merge as it can be misleading (we do not do in-place sort anymore, comment is suggesting that)

hubertp committed:
Update engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/SortNode.java

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

hubertp committed:
Disable a test that is different on Windows, as expected
